sqlglot/dialects/snowflake.py

```diff
@@ -1427,3 +1426,36 @@ def arrayagg_sql(self, expression: exp.ArrayAgg) -> str:
                 expr_sql = self.sql(exp.WithinGroup(this=expr_sql, expression=order))
 
             return expr_sql
+
+        def array_sql(self, expression: exp.Array) -> str:
+            expressions = expression.expressions
+
+            first_expr = seq_get(expressions, 0)
+            if isinstance(first_expr, exp.Select):
+                # SELECT AS STRUCT foo AS alias_foo -> ARRAY_AGG(OBJECT_CONSTRUCT('alias_foo', foo))
+                kind = first_expr.args.get("kind")
+                if kind and kind.upper() == "STRUCT":
+                    object_construct_args = []
+                    for expr in first_expr.expressions:
+                        if isinstance(expr, exp.Alias):
+                            # SELECT AS STRUCT foo AS alias_foo -> OBJECT_CONSTRUCT('alias_foo', foo)
+                            name = expr.this.name
+                            alias = f"'{expr.alias}'"
+                        else:
+                            # SELECT AS STRUCT foo -> OBJECT_CONSTRUCT('foo', foo)
+                            name = expr.name
+                            alias = f"'{name}'"
+
+                        object_construct_args.extend([alias, name])
+
+                    array_agg = exp.ArrayAgg(
+                        this=_build_object_construct(args=object_construct_args)
+                    )
+                    select = exp.select(array_agg, dialect=self.dialect)
+                    from_ = first_expr.args.get("from")
+                    if from_:
+                        select = select.from_(from_, dialect=self.dialect)
```

Comment:
What about other modifiers?

Reply:
Yes, that seems safer, otherwise you could silently drop query modifiers in the future, etc. May also wanna double-check whether `kind` needs to be switched off for Snowflake, I don't recall it being supported, but may be wrong.